### DISCUSSION TITLE
PCM output?

### DISCUSSION THREAD

**Discussion by Inssomniak** [pschatzmann/arduino-audio-tools#1682] [Q&A] [answered]
I’ve gone through the project and didn’t find an answer. I’m looking for a way to run an i2s stream from a microphone to PCM input for a cellular module.  Is there a part of this project that can do that?

**Comment by pschatzmann** [comment, marked as answer]
The simplest solution is just to [output the audio](https://github.com/pschatzmann/arduino-audio-tools/wiki/Audio-Input-and-Output) via the internal DAC (using the AnalogAudioStream).

**Comment by Inssomniak** [comment]
Wouldn’t that just net me an analog stream? I sadly can’t pipe that into the module. Esp32 of some variant is my target architecture.

**Reply by pschatzmann** [reply]
So you would need to provide some more information about your module. They usually provide
- audio pcm output
- microphone/audio pcm input 
- networking api

I assumed you wanted to use the audio functionality....

**Comment by Inssomniak** [comment]
It’s a sim7600 module. A Google of sim7600 Hardware design guide in Page 37 is the PCM hardware interface design.  I can’t cut and paste text at this time and can’t post a link. It samples at 8k or 16k depending on its configuration.

**Comment by pschatzmann** [comment]
This looks like an I2S API to me: did you try if this is working ?
Otherwise you need a codec chip if you want to process an analog input.

**Comment by Inssomniak** [comment]
Not yet, but after some more digging you might be right. Documents never mention i2s but it would make the most sense in this day and age. I’ll try and report back!On Aug 30, 2024, at 2:50 AM, Phil Schatzmann ***@***.***> wrote:﻿
This looks like an I2S API to me: did you try if this is working ?

**Comment by Inssomniak** [comment]
I did some research on this and wanted to run it by you to see if I’m just beating dead horses.. My module is actually expecting an external codec specifically the NAU8810 mono codec. Its firmware is written to control this codec via SPI, and it outputs i2s or PCM according to the data sheet. I only care about the mic input. 

When I look at the data sheet for its timings it seems that it’s the same for either i2s or PCM mode except where the MSB is. Ultimately I’d like to emulate this codec with an esp32 i2s peripheral, and input a MEMS mic into the esp32 via the same or a separate i2s peripheral and back out so I can possibly use audio detection to automatically tell the module to send an SMS, gain, equalize, etc. your project seems perfect fit minus this hang up. 

Expected:
Line Interface Format: Linear (fixed)
Data Length: Fixed 16 bits
PCM Clock/Sync Source: Master mode (fixed)
PCM Clock rate: 2048khz or 4096khz 
PCM Sync format: Short (fixed)
Data Ordering: MSB
